File: demoCode/taskDemo/taskDemo.c
#include <omp.h>
#include <stdio.h>
#include <unistd.h>

void foo(){
	printf("inside function foo, using thread number %d\n", omp_get_thread_num());
}

void bar(){
	printf("inside function bar and sleeping for 2 seconds, using thread number %d\n", omp_get_thread_num());
	sleep(2);
  printf("exiting bar\n");
}

int main(){
  printf("about to open a parallel region\n");
  printf("waiting for 2 seconds");
  sleep(2);

  double x = 10;
  // printf("x=%f\n", x);

  #pragma omp parallel firstprivate(x)
  {
    // printf("x=%f\n", x);
    printf("inside parallel region, using thread number %d. OMP makes no guarantee of ordering\n", omp_get_thread_num());
    sleep(2);

    #pragma omp single
    {
      printf("++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n");
      printf("inside single region. only one thread allowed to do this part\n");
      printf("no guarantee as to which one it is. this time it's %d\n", omp_get_thread_num());
      printf("spawning tasks: \n");

      #pragma omp task
      bar();

      #pragma omp task
      foo();

      sleep(3);
      printf("notice the order above. we began bar, and then immediately began foo\n");
      printf("we can do the same again, but force bar() to complete before spawning foo\n");
      printf("++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n");
    }

    #pragma omp master
    {
      printf("this time we use the master thread. it has number %d \n", omp_get_thread_num());
      printf("++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n");

      #pragma omp task
      bar();

      sleep(1);
      printf("waiting for bar to complete before continuing\n");
      #pragma omp taskwait

      #pragma omp task
      foo();
    }

  }
}

File: omp/examples/tasks/taskDemo.f90
PROGRAM taskDemo
   USE omp_lib
   IMPLICIT NONE

   DOUBLE PRECISION :: x = 10.0

   WRITE(*,*) 'about to open a parallel region'
   WRITE(*,*) 'waiting for 2 seconds'
   CALL sleep(2)

   !$OMP PARALLEL FIRSTPRIVATE(x)
   WRITE(*,'(A,I0,A)') 'inside parallel region, using thread number ', omp_get_thread_num(), &
                       '. OMP makes no guarantee of ordering'
   CALL sleep(2)

   !$OMP SINGLE
   WRITE(*,*) '+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++'
   WRITE(*,*) 'inside single region. only one thread allowed to do this part'
   WRITE(*,'(A,I0)') 'no guarantee as to which one it is. this time it''s ', omp_get_thread_num()
   WRITE(*,*) 'spawning tasks: '

   !$OMP TASK
   CALL bar()
   !$OMP END TASK

   !$OMP TASK  
   CALL foo()
   !$OMP END TASK

   CALL sleep(3)
   WRITE(*,*) 'notice the order above. we began bar, and then immediately began foo'
   WRITE(*,*) 'we can do the same again, but force bar() to complete before spawning foo'
   WRITE(*,*) '+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++'
   !$OMP END SINGLE

   !$OMP MASTER
   WRITE(*,'(A,I0)') 'this time we use the master thread. it has number ', omp_get_thread_num()
   WRITE(*,*) '+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++'

   !$OMP TASK
   CALL bar()
   !$OMP END TASK

   CALL sleep(1)
   WRITE(*,*) 'waiting for bar to complete before continuing'
   !$OMP TASKWAIT

   !$OMP TASK
   CALL foo()
   !$OMP END TASK
   !$OMP END MASTER

   !$OMP END PARALLEL

CONTAINS

   SUBROUTINE foo()
      USE omp_lib
      WRITE(*,'(A,I0)') 'inside function foo, using thread number ', omp_get_thread_num()
   END SUBROUTINE foo

   SUBROUTINE bar()
      USE omp_lib
      WRITE(*,'(A,I0)') 'inside function bar and sleeping for 2 seconds, using thread number ', &
                        omp_get_thread_num()
      CALL sleep(2)
      WRITE(*,*) 'exiting bar'
   END SUBROUTINE bar

END PROGRAM taskDemo